keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/AgpVersionHandler.kt

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2020 Slack Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// TODO Can't use the newer `com.android.Version` until AGP 3.6.0+
+@file:Suppress("deprecation")
+package com.slack.keeper
+
+import com.android.build.api.transform.Transform
+import com.android.build.gradle.internal.pipeline.TransformTask
+import com.android.builder.model.Version.ANDROID_GRADLE_PLUGIN_VERSION
+import com.google.auto.service.AutoService
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.Directory
+import org.gradle.api.provider.Provider
+import org.gradle.kotlin.dsl.withType
+import org.gradle.util.VersionNumber
+import java.lang.reflect.Field
+import java.util.Locale
+import kotlin.reflect.KProperty1
+import kotlin.reflect.full.memberProperties
+
+internal val AGP_VERSION get() = VersionNumber.parse(ANDROID_GRADLE_PLUGIN_VERSION)
+
+interface AgpVersionHandler {
+
+  companion object {
+    fun getInstance(): AgpVersionHandler = lazyInstance
+
+    private val lazyInstance by lazy {
+      createHandler()
+    }
+
+    /** Creates a new patcher for the given environment. */
+    private fun createHandler(): AgpVersionHandler {
+      /** Known patchers, listed in order of preference. */
+      // TODO(zsweers) properly use a ServiceLoader for these?
+      val handlers = listOf(Agp35xPatcher(), Agp36xPatcher())
+
+      val baseVersion = AGP_VERSION.baseVersion
+      return handlers.filter { baseVersion >= it.minVersion }
+          .maxBy(AgpVersionHandler::minVersion)
+          ?: error(
+              "$TAG No applicable proguard task patchers found for this version of AGP ($AGP_VERSION). Please file a bug report with your AGP version")
+    }
+  }
+
+  /** Minimum AGP version for this patcher. */
+  val minVersion: VersionNumber
+
+  /** Returns the expected minifier name (i.e. 'R8' or 'Proguard') for the givven [project]. */
+  fun expectedMinifier(project: Project): String
```

Comment:
Not: why expected? Is there an unexpected one?

Reply:
There could be, this is intentionally open-ended and stringly typed since there's other tools out there (redex, etc). We can always tweak it in the future since it's not a stable API :)